frontend/src/lib/utils.ts

```diff
@@ -203,3 +203,49 @@ export function labelVisibilityProtoValueToEnum(
       return LabelVisibilityOptions.Visible
   }
 }
+
+/**
+ * Tries to get an IFrame in which Streamlit app is embedded on Cloud deployments.
+ * It assumes iframe has title="streamlitApp", iterates over IFrames,
+ * and looks which IFrame contains div with stAppId value, otherwise returns first found iFrame or null.
+ */
+export function getIFrameEnclosingApp(
+  stAppId: string
+): HTMLIFrameElement | null {
+  if (!isInChildFrame()) {
+    return null
+  }
+  const qsStreamlitAppStr = 'iframe[title="streamlitApp"]'
+  let qs = window.document.querySelectorAll(
+    qsStreamlitAppStr
+  ) as NodeListOf<HTMLIFrameElement>
+  if (qs.length < 1 && window.parent) {
+    qs = window.parent.document.querySelectorAll(qsStreamlitAppStr)
+  }
```

Comment:
Ah shoot, I just realized this won't work if we need to reach outside the iframe because that will be disallowed in most deployments (specifically, browsers won't allow this to happen when the iframe is from a different origin from the host)

I also wonder if there's a more straightforward way to do this given that what we're really trying to do here is return this app's own `contentWindow`, which I would hope is more easily accessible from the context of the iframe itself rather than having to reach out into the parent and find the `contentWindow` that we're already in.

Since this code is executing from within the context of the iframe, I wonder if just calling `window` actually always does the right thing here since either:
* There's no enclosing iframe, so we can just open the browser's print dialog.
* There's an enclosing iframe, but calling `window` will execute in the context of the current iframe, so we actually won't have to worry about picking up changes from the parent or from other iframes in the enclosing page.

If that's the case, then `printCallback` in `App.tsx` would really only ever have to call `window.print`, and all of this extra code will have turned out to be unnecessary 😕

Reply:
@vdonato but calling just `window.print` sometimes printed only 1 page, instead of all of them, I got your worries this is a hacky solution, however it works, it also works on the Streamlit Cloud, so to improve it we can really just put it in `try catch` if you are worried that on some deployments it won't work

Reply:
Phew, crisis averted. I think this should be fine as long as we have error handling around the case where the browser prevents us from inspecting the parent frame